File: JasSwift/Sources/Algorithms/FitCurve.swift
import Foundation

/// A fitted cubic Bezier segment with endpoints and control points.
struct FitSegment {
    let p1x: Double, p1y: Double
    let c1x: Double, c1y: Double
    let c2x: Double, c2y: Double
    let p2x: Double, p2y: Double
}

/// Fit a cubic Bezier spline to a sequence of 2D points using the
/// Schneider algorithm (Graphics Gems I, 1990).
func fitCurve(points: [(Double, Double)], error: Double) -> [FitSegment] {
    guard points.count >= 2 else { return [] }
    let d = points
    let tHat1 = leftTangent(d, 0)
    let tHat2 = rightTangent(d, d.count - 1)
    var result: [FitSegment] = []
    fitCubic(d, 0, d.count - 1, tHat1, tHat2, error, &result)
    return result
}

private let maxIterations = 4

private func fitCubic(
    _ d: [(Double, Double)], _ first: Int, _ last: Int,
    _ tHat1: (Double, Double), _ tHat2: (Double, Double),
    _ error: Double, _ result: inout [FitSegment]
) {
    let nPts = last - first + 1

    if nPts == 2 {
        let dist = vdist(d[first], d[last]) / 3.0
        result.append(FitSegment(
            p1x: d[first].0, p1y: d[first].1,
            c1x: d[first].0 + tHat1.0 * dist, c1y: d[first].1 + tHat1.1 * dist,
            c2x: d[last].0 + tHat2.0 * dist, c2y: d[last].1 + tHat2.1 * dist,
            p2x: d[last].0, p2y: d[last].1
        ))
        return
    }

    var u = chordLengthParameterize(d, first, last)
    var bezCurve = generateBezier(d, first, last, u, tHat1, tHat2)
    var (maxError, splitPoint) = computeMaxError(d, first, last, bezCurve, u)

    if maxError < error {
        result.append(bezCurve)
        return
    }

    let iterationError = error * error
    if maxError < iterationError {
        for _ in 0..<maxIterations {
            let uPrime = reparameterize(d, first, last, u, bezCurve)
            bezCurve = generateBezier(d, first, last, uPrime, tHat1, tHat2)
            (maxError, splitPoint) = computeMaxError(d, first, last, bezCurve, uPrime)
            if maxError < error {
                result.append(bezCurve)
                return
            }
            u = uPrime
        }
    }

    let tHatCenter = centerTangent(d, splitPoint)
    fitCubic(d, first, splitPoint, tHat1, tHatCenter, error, &result)
    fitCubic(d, splitPoint, last, (-tHatCenter.0, -tHatCenter.1), tHat2, error, &result)
}

private func generateBezier(
    _ d: [(Double, Double)], _ first: Int, _ last: Int,
    _ uPrime: [Double],
    _ tHat1: (Double, Double), _ tHat2: (Double, Double)
) -> FitSegment {
    let nPts = last - first + 1

    let A: [((Double, Double), (Double, Double))] = (0..<nPts).map { i in
        (vscale(tHat1, b1(uPrime[i])), vscale(tHat2, b2(uPrime[i])))
    }

    var C = [[0.0, 0.0], [0.0, 0.0]]
    var X = [0.0, 0.0]

    for i in 0..<nPts {
        C[0][0] += vdot(A[i].0, A[i].0)
        C[0][1] += vdot(A[i].0, A[i].1)
        C[1][0] = C[0][1]
        C[1][1] += vdot(A[i].1, A[i].1)
        let tmp = vsub(
            d[first + i],
            vadd(vscale(d[first], b0(uPrime[i])),
                 vadd(vscale(d[first], b1(uPrime[i])),
                      vadd(vscale(d[last], b2(uPrime[i])),
                           vscale(d[last], b3(uPrime[i])))))
        )
        X[0] += vdot(A[i].0, tmp)
        X[1] += vdot(A[i].1, tmp)
    }

    let det_C0_C1 = C[0][0] * C[1][1] - C[1][0] * C[0][1]
    let det_C0_X = C[0][0] * X[1] - C[1][0] * X[0]
    let det_X_C1 = X[0] * C[1][1] - X[1] * C[0][1]

    let alpha_l = det_C0_C1 == 0 ? 0.0 : det_X_C1 / det_C0_C1
    let alpha_r = det_C0_C1 == 0 ? 0.0 : det_C0_X / det_C0_C1

    let segLength = vdist(d[first], d[last])
    let epsilon = 1.0e-6 * segLength

    if alpha_l < epsilon || alpha_r < epsilon {
        let dist = segLength / 3.0
        return FitSegment(
            p1x: d[first].0, p1y: d[first].1,
            c1x: d[first].0 + tHat1.0 * dist, c1y: d[first].1 + tHat1.1 * dist,
            c2x: d[last].0 + tHat2.0 * dist, c2y: d[last].1 + tHat2.1 * dist,
            p2x: d[last].0, p2y: d[last].1
        )
    }

    return FitSegment(
        p1x: d[first].0, p1y: d[first].1,
        c1x: d[first].0 + tHat1.0 * alpha_l, c1y: d[first].1 + tHat1.1 * alpha_l,
        c2x: d[last].0 + tHat2.0 * alpha_r, c2y: d[last].1 + tHat2.1 * alpha_r,
        p2x: d[last].0, p2y: d[last].1
    )
}

private func reparameterize(
    _ d: [(Double, Double)], _ first: Int, _ last: Int,
    _ u: [Double], _ bez: FitSegment
) -> [Double] {
    let pts = [(bez.p1x, bez.p1y), (bez.c1x, bez.c1y),
               (bez.c2x, bez.c2y), (bez.p2x, bez.p2y)]
    return (first...last).map { i in
        newtonRaphson(pts, d[i], u[i - first])
    }
}

private func newtonRaphson(
    _ Q: [(Double, Double)], _ P: (Double, Double), _ u: Double
) -> Double {
    let Q_u = bezierII(3, Q, u)
    let Q1 = [
        ((Q[1].0 - Q[0].0) * 3, (Q[1].1 - Q[0].1) * 3),
        ((Q[2].0 - Q[1].0) * 3, (Q[2].1 - Q[1].1) * 3),
        ((Q[3].0 - Q[2].0) * 3, (Q[3].1 - Q[2].1) * 3),
    ]
    let Q2 = [
        ((Q1[1].0 - Q1[0].0) * 2, (Q1[1].1 - Q1[0].1) * 2),
        ((Q1[2].0 - Q1[1].0) * 2, (Q1[2].1 - Q1[1].1) * 2),
    ]
    let Q1_u = bezierII(2, Q1, u)
    let Q2_u = bezierII(1, Q2, u)

    let numerator = (Q_u.0 - P.0) * Q1_u.0 + (Q_u.1 - P.1) * Q1_u.1
    let denominator = Q1_u.0 * Q1_u.0 + Q1_u.1 * Q1_u.1
        + (Q_u.0 - P.0) * Q2_u.0 + (Q_u.1 - P.1) * Q2_u.1

    if denominator == 0 { return u }
    return u - numerator / denominator
}

private func bezierII(_ degree: Int, _ V: [(Double, Double)], _ t: Double) -> (Double, Double) {
    var Vtemp = V
    for i in 1...degree {
        for j in 0...(degree - i) {
            Vtemp[j] = (
                (1.0 - t) * Vtemp[j].0 + t * Vtemp[j + 1].0,
                (1.0 - t) * Vtemp[j].1 + t * Vtemp[j + 1].1
            )
        }
    }
    return Vtemp[0]
}

private func computeMaxError(
    _ d: [(Double, Double)], _ first: Int, _ last: Int,
    _ bez: FitSegment, _ u: [Double]
) -> (Double, Int) {
    let pts = [(bez.p1x, bez.p1y), (bez.c1x, bez.c1y),
               (bez.c2x, bez.c2y), (bez.p2x, bez.p2y)]
    var splitPoint = (last - first + 1) / 2
    var maxDist = 0.0
    for i in (first + 1)..<last {
        let P = bezierII(3, pts, u[i - first])
        let dx = P.0 - d[i].0, dy = P.1 - d[i].1
        let dist = dx * dx + dy * dy
        if dist >= maxDist {
            maxDist = dist
            splitPoint = i
        }
    }
    return (maxDist, splitPoint)
}

private func chordLengthParameterize(
    _ d: [(Double, Double)], _ first: Int, _ last: Int
) -> [Double] {
    var u = [Double](repeating: 0, count: last - first + 1)
    for i in (first + 1)...last {
        u[i - first] = u[i - first - 1] + vdist(d[i], d[i - 1])
    }
    let total = u[last - first]
    if total > 0 {
        for i in (first + 1)...last {
            u[i - first] /= total
        }
    }
    return u
}

private func leftTangent(_ d: [(Double, Double)], _ end: Int) -> (Double, Double) {
    vnormalize(vsub(d[end + 1], d[end]))
}
private func rightTangent(_ d: [(Double, Double)], _ end: Int) -> (Double, Double) {
    vnormalize(vsub(d[end - 1], d[end]))
}
private func centerTangent(_ d: [(Double, Double)], _ center: Int) -> (Double, Double) {
    let v1 = vsub(d[center - 1], d[center])
    let v2 = vsub(d[center], d[center + 1])
    return vnormalize(((v1.0 + v2.0) / 2, (v1.1 + v2.1) / 2))
}

// Bernstein basis functions
private func b0(_ u: Double) -> Double { let t = 1 - u; return t * t * t }
private func b1(_ u: Double) -> Double { let t = 1 - u; return 3 * u * t * t }
private func b2(_ u: Double) -> Double { let t = 1 - u; return 3 * u * u * t }
private func b3(_ u: Double) -> Double { u * u * u }

// Vector helpers
private func vadd(_ a: (Double, Double), _ b: (Double, Double)) -> (Double, Double) {
    (a.0 + b.0, a.1 + b.1)
}
private func vsub(_ a: (Double, Double), _ b: (Double, Double)) -> (Double, Double) {
    (a.0 - b.0, a.1 - b.1)
}
private func vscale(_ v: (Double, Double), _ s: Double) -> (Double, Double) {
    (v.0 * s, v.1 * s)
}
private func vdot(_ a: (Double, Double), _ b: (Double, Double)) -> Double {
    a.0 * b.0 + a.1 * b.1
}
private func vdist(_ a: (Double, Double), _ b: (Double, Double)) -> Double {
    hypot(a.0 - b.0, a.1 - b.1)
}
private func vnormalize(_ v: (Double, Double)) -> (Double, Double) {
    let len = hypot(v.0, v.1)
    if len == 0 { return v }
    return (v.0 / len, v.1 / len)
}

File: jas_ocaml/lib/algorithms/fit_curve.ml
(** Bezier curve fitting using the Schneider algorithm.

    Fits a sequence of points to a piecewise cubic Bezier curve.
    Based on "An Algorithm for Automatically Fitting Digitized Curves"
    by Philip J. Schneider, Graphics Gems I, 1990. *)

type segment = {
  p1x: float; p1y: float;
  c1x: float; c1y: float;
  c2x: float; c2y: float;
  p2x: float; p2y: float;
}

(* Vector helpers *)
let vadd (ax, ay) (bx, by) = (ax +. bx, ay +. by)
let vsub (ax, ay) (bx, by) = (ax -. bx, ay -. by)
let vscale (vx, vy) s = (vx *. s, vy *. s)
let vdot (ax, ay) (bx, by) = ax *. bx +. ay *. by
let vdist (ax, ay) (bx, by) = hypot (ax -. bx) (ay -. by)
let vnormalize (vx, vy) =
  let len = hypot vx vy in
  if len = 0.0 then (vx, vy)
  else (vx /. len, vy /. len)

(* Bernstein basis functions *)
let b0 u = let t = 1.0 -. u in t *. t *. t
let b1 u = let t = 1.0 -. u in 3.0 *. u *. t *. t
let b2 u = let t = 1.0 -. u in 3.0 *. u *. u *. t
let b3 u = u *. u *. u

let max_iterations = 4

let left_tangent d idx = vnormalize (vsub d.(idx + 1) d.(idx))
let right_tangent d idx = vnormalize (vsub d.(idx - 1) d.(idx))
let center_tangent d idx =
  let v1 = vsub d.(idx - 1) d.(idx) in
  let v2 = vsub d.(idx) d.(idx + 1) in
  vnormalize ((fst v1 +. fst v2) /. 2.0, (snd v1 +. snd v2) /. 2.0)

let chord_length_parameterize d first last =
  let n = last - first + 1 in
  let u = Array.make n 0.0 in
  for i = first + 1 to last do
    u.(i - first) <- u.(i - first - 1) +. vdist d.(i) d.(i - 1)
  done;
  let total = u.(last - first) in
  if total > 0.0 then
    for i = first + 1 to last do
      u.(i - first) <- u.(i - first) /. total
    done;
  u

let bezier_ii degree v t =
  let vtemp = Array.copy v in
  for i = 1 to degree do
    for j = 0 to degree - i do
      let (x0, y0) = vtemp.(j) in
      let (x1, y1) = vtemp.(j + 1) in
      vtemp.(j) <- ((1.0 -. t) *. x0 +. t *. x1, (1.0 -. t) *. y0 +. t *. y1)
    done
  done;
  vtemp.(0)

let newton_raphson q p u =
  let q_u = bezier_ii 3 q u in
  let q1 = [|
    ((fst q.(1) -. fst q.(0)) *. 3.0, (snd q.(1) -. snd q.(0)) *. 3.0);
    ((fst q.(2) -. fst q.(1)) *. 3.0, (snd q.(2) -. snd q.(1)) *. 3.0);
    ((fst q.(3) -. fst q.(2)) *. 3.0, (snd q.(3) -. snd q.(2)) *. 3.0);
  |] in
  let q2 = [|
    ((fst q1.(1) -. fst q1.(0)) *. 2.0, (snd q1.(1) -. snd q1.(0)) *. 2.0);
    ((fst q1.(2) -. fst q1.(1)) *. 2.0, (snd q1.(2) -. snd q1.(1)) *. 2.0);
  |] in
  let q1_u = bezier_ii 2 q1 u in
  let q2_u = bezier_ii 1 q2 u in
  let numerator =
    (fst q_u -. fst p) *. fst q1_u +. (snd q_u -. snd p) *. snd q1_u
  in
  let denominator =
    fst q1_u *. fst q1_u +. snd q1_u *. snd q1_u
    +. (fst q_u -. fst p) *. fst q2_u
    +. (snd q_u -. snd p) *. snd q2_u
  in
  if denominator = 0.0 then u
  else u -. numerator /. denominator

let generate_bezier d first last u_prime that1 that2 =
  let n_pts = last - first + 1 in
  let a = Array.init n_pts (fun i ->
    (vscale that1 (b1 u_prime.(i)), vscale that2 (b2 u_prime.(i)))
  ) in
  let c = [| [| 0.0; 0.0 |]; [| 0.0; 0.0 |] |] in
  let x = [| 0.0; 0.0 |] in
  for i = 0 to n_pts - 1 do
    c.(0).(0) <- c.(0).(0) +. vdot (fst a.(i)) (fst a.(i));
    c.(0).(1) <- c.(0).(1) +. vdot (fst a.(i)) (snd a.(i));
    c.(1).(0) <- c.(0).(1);
    c.(1).(1) <- c.(1).(1) +. vdot (snd a.(i)) (snd a.(i));
    let tmp = vsub d.(first + i)
      (vadd (vscale d.(first) (b0 u_prime.(i)))
        (vadd (vscale d.(first) (b1 u_prime.(i)))
          (vadd (vscale d.(last) (b2 u_prime.(i)))
            (vscale d.(last) (b3 u_prime.(i)))))) in
    x.(0) <- x.(0) +. vdot (fst a.(i)) tmp;
    x.(1) <- x.(1) +. vdot (snd a.(i)) tmp
  done;
  let det_c0_c1 = c.(0).(0) *. c.(1).(1) -. c.(1).(0) *. c.(0).(1) in
  let det_c0_x = c.(0).(0) *. x.(1) -. c.(1).(0) *. x.(0) in
  let det_x_c1 = x.(0) *. c.(1).(1) -. x.(1) *. c.(0).(1) in
  let alpha_l = if det_c0_c1 = 0.0 then 0.0 else det_x_c1 /. det_c0_c1 in
  let alpha_r = if det_c0_c1 = 0.0 then 0.0 else det_c0_x /. det_c0_c1 in
  let seg_length = vdist d.(first) d.(last) in
  let epsilon = 1.0e-6 *. seg_length in
  if alpha_l < epsilon || alpha_r < epsilon then begin
    let dist = seg_length /. 3.0 in
    { p1x = fst d.(first); p1y = snd d.(first);
      c1x = fst d.(first) +. fst that1 *. dist;
      c1y = snd d.(first) +. snd that1 *. dist;
      c2x = fst d.(last) +. fst that2 *. dist;
      c2y = snd d.(last) +. snd that2 *. dist;
      p2x = fst d.(last); p2y = snd d.(last) }
  end else
    { p1x = fst d.(first); p1y = snd d.(first);
      c1x = fst d.(first) +. fst that1 *. alpha_l;
      c1y = snd d.(first) +. snd that1 *. alpha_l;
      c2x = fst d.(last) +. fst that2 *. alpha_r;
      c2y = snd d.(last) +. snd that2 *. alpha_r;
      p2x = fst d.(last); p2y = snd d.(last) }

let compute_max_error d first last bez u =
  let pts = [| (bez.p1x, bez.p1y); (bez.c1x, bez.c1y);
               (bez.c2x, bez.c2y); (bez.p2x, bez.p2y) |] in
  let split_point = ref ((last - first + 1) / 2) in
  let max_dist = ref 0.0 in
  for i = first + 1 to last - 1 do
    let p = bezier_ii 3 pts u.(i - first) in
    let dx = fst p -. fst d.(i) in
    let dy = snd p -. snd d.(i) in
    let dist = dx *. dx +. dy *. dy in
    if dist >= !max_dist then begin
      max_dist := dist;
      split_point := i
    end
  done;
  (!max_dist, !split_point)

let reparameterize d first last u bez =
  let pts = [| (bez.p1x, bez.p1y); (bez.c1x, bez.c1y);
               (bez.c2x, bez.c2y); (bez.p2x, bez.p2y) |] in
  Array.init (last - first + 1) (fun i ->
    newton_raphson pts d.(first + i) u.(i)
  )

let rec fit_cubic d first last that1 that2 error result =
  let n_pts = last - first + 1 in
  if n_pts = 2 then begin
    let dist = vdist d.(first) d.(last) /. 3.0 in
    let seg = {
      p1x = fst d.(first); p1y = snd d.(first);
      c1x = fst d.(first) +. fst that1 *. dist;
      c1y = snd d.(first) +. snd that1 *. dist;
      c2x = fst d.(last) +. fst that2 *. dist;
      c2y = snd d.(last) +. snd that2 *. dist;
      p2x = fst d.(last); p2y = snd d.(last);
    } in
    result := seg :: !result
  end else begin
    let u = chord_length_parameterize d first last in
    let bez = ref (generate_bezier d first last u that1 that2) in
    let (max_err, sp) = compute_max_error d first last !bez u in
    let max_error = ref max_err in
    let split_point = ref sp in
    if !max_error < error then
      result := !bez :: !result
    else begin
      let iteration_error = error *. error in
      let u_ref = ref u in
      if !max_error < iteration_error then begin
        let done_ = ref false in
        for _ = 0 to max_iterations - 1 do
          if not !done_ then begin
            let u_prime = reparameterize d first last !u_ref !bez in
            bez := generate_bezier d first last u_prime that1 that2;
            let (me, sp2) = compute_max_error d first last !bez u_prime in
            max_error := me;
            split_point := sp2;
            if !max_error < error then begin
              result := !bez :: !result;
              done_ := true
            end else
              u_ref := u_prime
          end
        done;
        if !done_ then ()
        else begin
          let tc = center_tangent d !split_point in
          fit_cubic d first !split_point that1 tc error result;
          let tc_neg = (-. fst tc, -. snd tc) in
          fit_cubic d !split_point last tc_neg that2 error result
        end
      end else begin
        let tc = center_tangent d !split_point in
        fit_cubic d first !split_point that1 tc error result;
        let tc_neg = (-. fst tc, -. snd tc) in
        fit_cubic d !split_point last tc_neg that2 error result
      end
    end
  end

let fit_curve (points : (float * float) list) (error : float) : segment list =
  match points with
  | [] | [_] -> []
  | _ ->
    let d = Array.of_list points in
    let n = Array.length d in
    let that1 = left_tangent d 0 in
    let that2 = right_tangent d (n - 1) in
    let result = ref [] in
    fit_cubic d 0 (n - 1) that1 that2 error result;
    List.rev !result
